src/useSubscription.ts

```diff
@@ -0,0 +1,86 @@
+import {
+  ApolloClient,
+  ApolloError,
+  OperationVariables,
+  SubscriptionOptions,
+} from 'apollo-client';
+import { DocumentNode } from 'graphql';
+import { useEffect, useRef, useState } from 'react';
+import { useApolloClient } from './ApolloContext';
+import { Omit, objToKey } from './utils';
+
+export type OnSubscriptionData<TData> = (
+  options: OnSubscriptionDataOptions<TData>
+) => any;
+
+export interface OnSubscriptionDataOptions<TData> {
+  client: ApolloClient<any>;
+  subscriptionData: SubscriptionHookResult<TData>;
+}
+
+export interface SubscriptionHookOptions<TData, TVariables>
+  extends Omit<SubscriptionOptions<TVariables>, 'query'> {
+  onSubscriptionData?: OnSubscriptionData<TData>;
+}
+
+export interface SubscriptionHookResult<TData> {
+  data?: TData;
+  error?: ApolloError;
+  loading: boolean;
+}
+
+export function useSubscription<TData = any, TVariables = OperationVariables>(
+  query: DocumentNode,
+  {
+    variables,
+    onSubscriptionData,
+    ...options
+  }: SubscriptionHookOptions<TData, TVariables> = {}
+): SubscriptionHookResult<TData> {
+  const client = useApolloClient();
+  const onSubscriptionDataRef = useRef<null | OnSubscriptionData<TData>>(null);
+  const [result, setResult] = useState<SubscriptionHookResult<TData>>({
+    loading: true,
+  });
+
+  if (onSubscriptionData) {
```

Comment:
I don't think this `if` is correct. In my opinion `onSubscriptionDataRef.current = onSubscriptionData;` should always be invoked. Otherwise, it's possible that the old `onSubscriptionData` callback will be invoked if it is replaced with `null`. E. g. please consider this snippet:

```typescript
const [aValue, setValue] = useState(1);
useSubscription(MY_SUBSCRIPTION, {
  onSubscriptionData: (aValue === 1) ? someCallback : null,
});
```

Reply:
Feels like edge case why would someone want conditionally set the callback while it's easier to check the condition inside that callback. Doesn't mean it shouldn't be fixed though :)